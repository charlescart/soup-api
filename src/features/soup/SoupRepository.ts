/* eslint-disable no-param-reassign */
/* eslint-disable no-continue */
/* eslint-disable no-console */
/* eslint-disable class-methods-use-this */
import soupDto from './validators/soupDto';
import summaryDto from './validators/summaryDto';

export default class SoupRepository {
  static Soup({ soup, searchWords }: soupDto): summaryDto {
    const summary: summaryDto = this.CreateSummary(searchWords);

    const numberRows: number = soup.length - 1;
    soup.forEach((row: string[], rowIndex: number) => {
      const numberColumns: number = row.length - 1;
      row.forEach((letter: string, columnIndex: number) => {
        letter = letter.toUpperCase();

        this.LookUpOneLetterWords(letter, summary);
        this.LookToTheRight(columnIndex, numberColumns, summary, row);
        this.LookAtDiagonal(rowIndex, columnIndex, summary, soup);
      });
    });

    return summary;
  }

  static LookUpOneLetterWords(letter: string, summary: summaryDto): summaryDto {
    const wordsSizeOne: string[] = Object.keys(summary)
      .filter((word) => word.length === 1);

    for (let i = 0; i < wordsSizeOne.length; i += 1) {
      const word: string = wordsSizeOne[i];
      if (letter === word) summary[word] += 1;
    }

    return summary;
  }

  static LookToTheRight(columnIndex: number, columns: number, summary: summaryDto, row: string[]): summaryDto {
    const spaceToTheRight: number = columns - columnIndex;
    const words: string[] = Object.keys(summary)
      .filter((word) => (word.length - 1) >= 1 && (word.length - 1) <= spaceToTheRight);

    for (let i = 0; i < words.length; i += 1) {
      const word: string = words[i];
      const limit: number = (columnIndex + spaceToTheRight);
      const phrase: string = row
        .filter((item, index) => index >= columnIndex && index <= limit)
        .join('')
        .toUpperCase();

      if (word === phrase.substr(0, word.length)) summary[word] += 1;
    }

    return summary;
  }

  static LookAtDiagonal(rowIndex: number, columnIndex: number, summary: summaryDto, soup: string[][]): summaryDto {
    const numberRows: number = soup.length;
    let numberColumns: number = soup[rowIndex].length;
    let phraseDiagonal = '';

    for (let i = rowIndex; i < numberColumns; i += 1) {
      for (let j = columnIndex; j < numberColumns && i < numberRows; j += 1) {
        phraseDiagonal += soup[i][j].toUpperCase();
        i += 1;
        if (soup[i]) numberColumns = soup[i].length;
      }
      break;
    }

    if (phraseDiagonal.length === 0) return summary;

    const words: string[] = Object.keys(summary)
      .filter((word) => word.length > 1 && word.length <= phraseDiagonal.length);

    words.forEach((word) => {
      if (word === phraseDiagonal.substr(0, word.length)) summary[word] += 1;
    });

    return summary;
  }

  static CreateSummary(words: string[]): summaryDto {
    const summary: summaryDto = {};

    for (let i = 0; i < words.length; i += 1) {
      const word: string = words[i];
      if (word.length > 0) summary[word.toUpperCase()] = 0;
    }

    return summary;
  }
}
