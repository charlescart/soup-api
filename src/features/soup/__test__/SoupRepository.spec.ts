import SoupRepository from '../SoupRepository';
import Summary from '../validators/summaryDto';

describe('SoupRepository', () => {
  it('CreateSummary', () => {
    const words: string[] = ["oiE", "ixE", "Ix", "i", "O", "awds", "e4", "", "oi", "eX"];
    let summary: Summary = SoupRepository.CreateSummary(words);
    expect(summary).toMatchSnapshot();
  });

  it('LookUpOneLetterWords', () => {
    const words: string[] = ['o', 'I', 'e', 'X', '4', 'f', 'k', 'p', '1', '2', '3'];
    let summary: Summary = SoupRepository.CreateSummary(words);
    const soup: string[][] = [
      ["o", "I", "E"],
      ["I", "i", "X", "4"],
      ["E", "x", "E"],
      ["E", "X"],
      ["e"],
      [""]
    ];

    soup.forEach((row: string[]) => {
      row.forEach((letter: string) => {
        letter = letter.toUpperCase();
        SoupRepository.LookUpOneLetterWords(letter, summary);
      });
    });

    expect(summary).toMatchSnapshot();
  });

  it('LookToTheRight', () => {
    const words: string[] = ["oiE", "Oi", "ie", "ii", "eXe", "EX", "XE", "X4", "IIX4", "", "palabraSuperGrande"];
    let summary: Summary = SoupRepository.CreateSummary(words);
    const soup: string[][] = [
      ["o", "I", "E"],
      ["I", "i", "X", "4"],
      ["E", "x", "E"],
      ["E", "X"],
      ["e"],
      [""]
    ];

    soup.forEach((row: string[], rowIndex: number) => {
      const numberColumns: number = row.length - 1;
      row.forEach((letter: string, columnIndex: number) => {
        letter = letter.toUpperCase();
        SoupRepository.LookToTheRight(columnIndex, numberColumns, summary, row);
      });
    });

    expect(summary).toMatchSnapshot();
  });

  it('LookAtDiagonal', () => {
    const words: string[] = ["palabraSuperGrande", "oie", "OI", "iX", "E4", "Ex", ""];
    let summary: Summary = SoupRepository.CreateSummary(words);
    const soup: string[][] = [
      ["o", "I", "E"],
      ["I", "i", "X", "4"],
      ["E", "x", "E"],
      ["E", "X"],
      ["e"],
      [""]
    ];

    soup.forEach((row: string[], rowIndex: number) => {
      const numberColumns: number = row.length - 1;
      row.forEach((letter: string, columnIndex: number) => {
        letter = letter.toUpperCase();
        SoupRepository.LookAtDiagonal(rowIndex, columnIndex, summary, soup);
      });
    });

    expect(summary).toMatchSnapshot();
  });
});