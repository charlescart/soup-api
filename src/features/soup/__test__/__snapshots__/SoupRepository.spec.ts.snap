// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SoupRepository CreateSummary 1`] = `
Object {
  "AWDS": 0,
  "E4": 0,
  "EX": 0,
  "I": 0,
  "IX": 0,
  "IXE": 0,
  "O": 0,
  "OI": 0,
  "OIE": 0,
}
`;

exports[`SoupRepository LookAtDiagonal 1`] = `
Object {
  "E4": 1,
  "EX": 1,
  "IX": 2,
  "OI": 1,
  "OIE": 1,
  "PALABRASUPERGRANDE": 0,
}
`;

exports[`SoupRepository LookToTheRight 1`] = `
Object {
  "EX": 2,
  "EXE": 1,
  "IE": 1,
  "II": 1,
  "IIX4": 1,
  "OI": 1,
  "OIE": 1,
  "PALABRASUPERGRANDE": 0,
  "X4": 1,
  "XE": 1,
}
`;

exports[`SoupRepository LookUpOneLetterWords 1`] = `
Object {
  "1": 0,
  "2": 0,
  "3": 0,
  "4": 1,
  "E": 5,
  "F": 0,
  "I": 3,
  "K": 0,
  "O": 1,
  "P": 0,
  "X": 3,
}
`;
